File: day4/first/main.py
sum = 0

with open("../input.txt") as file:
    while (line := file.readline()):
        parts = line.split("|")
        winners = sorted(map(int, parts[0][parts[0].index(":")+1:].strip().split()))
        nums = sorted(map(int, parts[1].strip().split()))

        win_count = 0
        i, j = 0, 0
        while i < len(winners) and j < len(nums):
            if winners[i] == nums[j]:
                win_count += 1
                i += 1
                j += 1
                continue
            if winners[i] > nums[j]:
                j += 1
            else: 
                i += 1
        
        sum += int(2 ** (win_count-1))

print(sum)

File: day4/first/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

func main() {
	input, err := os.Open("../input.txt")
	if err != nil {
		os.Exit(1)
	}
	scanner := bufio.NewScanner(input)

	sum := 0
	for scanner.Scan() {
		card := scanner.Text()
		parts := strings.Split(card, "|")
		winners := atoi(strings.Fields(strings.TrimSpace(parts[0][strings.Index(parts[0], ":")+1:])))
		nums := atoi(strings.Fields(strings.TrimSpace(parts[1])))
		slices.Sort(winners)
		slices.Sort(nums)

		winCount := 0
		for i, j := 0, 0; i < len(winners) && j < len(nums); {
			if winners[i] == nums[j] {
				winCount++
				i++
				j++
				continue
			}
			if winners[i] > nums[j] {
				j++
			} else {
				i++
			}
		}
		sum += int(math.Pow(2, float64(winCount)-1))
	}
	fmt.Println(sum)
}

func atoi(slice []string) []int {
	intSlice := []int{}
	for _, e := range slice {
		num, _ := strconv.Atoi(e)
		intSlice = append(intSlice, num)
	}
	return intSlice
}
